tests/test_drop_counters.py

```diff
@@ -0,0 +1,526 @@
+        self.setup_db(dvs)
+
+        asic_state_table = swsscommon.Table(self.asic_db, ASIC_STATE_TABLE)
+        flex_counter_table = swsscommon.Table(self.flex_db, FLEX_COUNTER_TABLE)
+
+        name = 'ADD_TEST'
+        reason1 = 'L3_ANY'
+        reason2 = 'ACL_ANY'
+
+        self.create_drop_counter(name, SWITCH_INGRESS_DROPS)
+        self.add_drop_reason(name, reason1)
+        time.sleep(3)
+
+        # Quick sanity check to ensure the counter has been created
+        assert len(asic_state_table.getKeys()) == 1
+        assert len(flex_counter_table.getKeys()) > 0
+
+        self.remove_drop_reason(name, reason2)
+        time.sleep(3)
+
+        # Check that the ASIC state is correct
+        asic_keys = asic_state_table.getKeys()
+        assert len(asic_keys) == 1
+        assert self.checkASICState(asic_keys[0], ASIC_COUNTER_SWITCH_IN_TYPE, [reason1])
+
+        # cleanup
+        self.delete_drop_counter(name)
+        self.remove_drop_reason(name, reason1)
+
+    def test_removeInvalidDropReason(self, dvs, testlog):
+        self.setup_db(dvs)
+
+        asic_state_table = swsscommon.Table(self.asic_db, ASIC_STATE_TABLE)
+        flex_counter_table = swsscommon.Table(self.flex_db, FLEX_COUNTER_TABLE)
+
+        name = 'ADD_TEST'
+        reason1 = 'L3_ANY'
+        bogus_reason = 'LIVE_LAUGH_LOVE'
+
+        self.create_drop_counter(name, SWITCH_INGRESS_DROPS)
+        self.add_drop_reason(name, reason1)
+        time.sleep(3)
+
+        # Quick sanity check to ensure the counter has been created
+        assert len(asic_state_table.getKeys()) == 1
+        assert len(flex_counter_table.getKeys()) > 0
+
+        self.remove_drop_reason(name, bogus_reason)
+        time.sleep(3)
+
+        # Check that the ASIC state is correct
+        asic_keys = asic_state_table.getKeys()
+        assert len(asic_keys) == 1
+        assert self.checkASICState(asic_keys[0], ASIC_COUNTER_SWITCH_IN_TYPE, [reason1])
+
+        # cleanup
+        self.delete_drop_counter(name)
+        self.remove_drop_reason(name, reason1)
+
+    def test_createAndDeleteMultipleCounters(self, dvs, testlog):
+        self.setup_db(dvs)
+
+        asic_state_table = swsscommon.Table(self.asic_db, ASIC_STATE_TABLE)
+        flex_counter_table = swsscommon.Table(self.flex_db, FLEX_COUNTER_TABLE)
+
+        name1 = 'DEBUG_0'
+        reason1 = 'L3_ANY'
+
+        name2 = 'DEBUG_1'
+        reason2 = 'ACL_ANY'
+
+        self.create_drop_counter(name1, PORT_INGRESS_DROPS)
+        self.add_drop_reason(name1, reason1)
+        self.create_drop_counter(name2, PORT_INGRESS_DROPS)
+        self.add_drop_reason(name2, reason2)
+        time.sleep(5)
+
+        assert len(flex_counter_table.getKeys()) > 0
+
+        asic_keys = asic_state_table.getKeys()
+        assert len(asic_keys) == 2
+        for key in asic_keys:
+            assert (self.checkASICState(key, ASIC_COUNTER_PORT_IN_TYPE, [reason1]) or self.checkASICState(key, ASIC_COUNTER_PORT_IN_TYPE, [reason2]))
```

Comment:
I think your code can deal with either create_counter(), then create_reason(), or create_reason(), then create_counter(). But your test seems only testing the first ordering?

Reply:
There is a "createAndRemoveDropCounterBasic" which does the first ordering and "createAndRemoveDropCounterReversed" which does the second ordering. I've added comments to clarify this. :)